File: core/src/nl/t64/game/rpg/components/party/InventoryContainer.java
package nl.t64.game.rpg.components.party;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;


public class InventoryContainer {

    private static final int NUMBER_OF_SLOTS = 66;

    private final List<InventoryItem> inventory;

    public InventoryContainer() {
        this(NUMBER_OF_SLOTS);
    }

    public InventoryContainer(int numberOfSlots) {
        this.inventory = new ArrayList<>(Collections.nCopies(numberOfSlots, null));
    }

    public List<InventoryItem> getAllFilledSlots() {
        return inventory.stream()
                        .filter(Objects::nonNull)
                        .collect(Collectors.toList());
    }

    public int getAmountOfItemAt(int index) {
        return getItemAt(index).map(item -> item.amount)
                               .orElse(0);
    }

    public void incrementAmountAt(int index, int amount) {
        getItemAt(index).ifPresentOrElse(inventoryItem -> inventoryItem.increaseAmountWith(amount),
                                         throwException("There is no item to increment amount."));
    }

    public void decrementAmountAt(int index, int amount) {
        getItemAt(index).ifPresentOrElse(inventoryItem -> inventoryItem.decreaseAmountWith(amount),
                                         throwException("There is no item to decrement amount."));
    }

    public Optional<InventoryItem> getItemAt(int index) {
        return Optional.ofNullable(inventory.get(index));
    }

    public void autoSetItem(InventoryItem newItem) {
        if (newItem.isStackable()) {
            addResource(newItem);
        } else {
            addItemAtEmptySlot(newItem);
        }
    }

    public void autoRemoveItem(String itemId, int amount) {
        if (!hasEnoughOfItem(itemId, amount)) {
            throw new IllegalStateException("Cannot remove this resource from Inventory.");
        }
        for (Map.Entry<Integer, Integer> set : findAllSlotsWithAmountOfItem(itemId).entrySet()) {
            if (amount == 0) {
                break;
            } else if (amount >= set.getValue()) {
                amount -= set.getValue();
                clearItemAt(set.getKey());
            } else {
                decrementAmountAt(set.getKey(), amount);
                amount = 0;
            }
        }
    }

    public void clearItemAt(int index) {
        forceSetItemAt(index, null);
    }

    public void forceSetItemAt(int index, InventoryItem newItem) {
        inventory.set(index, newItem);
    }

    public int getSize() {
        return inventory.size();
    }

    public boolean isEmpty() {
        return inventory.stream().noneMatch(Objects::nonNull);
    }

    public void sort() {
        new InventoryStacksMerger(this).searchAll();
        Comparator<InventoryItem> comparing = Comparator.comparing(this::getInventoryGroup)
                                                        .thenComparing(this::getSort);
        inventory.sort(comparing);
    }

    public boolean contains(Map<String, Integer> items) {
        return items.entrySet()
                    .stream()
                    .allMatch(item -> hasEnoughOfItem(item.getKey(), item.getValue()));
    }

    public boolean hasEnoughOfItem(String itemId, int amount) {
        return getTotalOfItem(itemId) >= amount;
    }

    public boolean hasRoomForResource(String itemId) {
        return findFirstSlotWithItem(itemId).isPresent()
               || findFirstEmptySlot().isPresent();
    }

    int getLastIndex() {
        return getSize() - 1;
    }

    boolean contains(String itemId) {
        return inventory.stream()
                        .filter(Objects::nonNull)
                        .anyMatch(item -> item.hasSameIdAs(itemId));
    }

    int getTotalOfItem(String itemId) {
        return inventory.stream()
                        .filter(Objects::nonNull)
                        .filter(item -> item.hasSameIdAs(itemId))
                        .mapToInt(item -> item.amount)
                        .sum();
    }

    long getNumberOfFilledSlots() {
        return inventory.stream()
                        .filter(Objects::nonNull)
                        .count();
    }

    private void addResource(InventoryItem newItem) {
        getItem(newItem.id).ifPresentOrElse(inventoryItem -> inventoryItem.increaseAmountWith(newItem),
                                            () -> addItemAtEmptySlot(newItem)
        );
    }

    private void addItemAtEmptySlot(InventoryItem newItem) {
        findFirstEmptySlot().ifPresentOrElse(index -> slotIsEmptySoSetItemAt(index, newItem),
                                             throwException("Inventory is full."));
    }

    private void slotIsEmptySoSetItemAt(int index, InventoryItem newItem) {
        forceSetItemAt(index, newItem);
    }

    private Optional<InventoryItem> getItem(String itemId) {
        return inventory.stream()
                        .filter(Objects::nonNull)
                        .filter(item -> item.hasSameIdAs(itemId))
                        .findFirst();
    }

    private Map<Integer, Integer> findAllSlotsWithAmountOfItem(String itemId) {
        Map<Integer, Integer> resourceMap = new HashMap<>();
        IntStream.range(0, getSize())
                 .forEach(i -> {
                     InventoryItem item = inventory.get(i);
                     if (item != null && item.id.equals(itemId)) {
                         resourceMap.put(i, item.amount);
                     }
                 });
        return resourceMap;
    }

    public Optional<Integer> findFirstSlotWithItem(String itemId) {
        return IntStream.range(0, getSize()).boxed()
                        .filter(index -> containsItemAt(index, itemId))
                        .findFirst();
    }

    public OptionalInt findFirstFilledSlot() {
        return IntStream.range(0, getSize())
                        .filter(index -> !isSlotEmpty(index))
                        .findFirst();
    }

    public Optional<Integer> findFirstEmptySlot() {
        return IntStream.range(0, getSize()).boxed()
                        .filter(this::isSlotEmpty)
                        .findFirst();
    }

    private boolean isSlotEmpty(int index) {
        return inventory.get(index) == null;
    }

    private boolean containsItemAt(int index, String itemId) {
        return getItemAt(index).map(inventoryItem -> inventoryItem.hasSameIdAs(itemId))
                               .orElse(false);
    }

    private InventoryGroup getInventoryGroup(InventoryItem inventoryItem) {
        if (inventoryItem == null) {
            return InventoryGroup.EMPTY;
        }
        return inventoryItem.group;
    }

    private int getSort(InventoryItem inventoryItem) {
        if (inventoryItem == null) {
            return 0;
        }
        return inventoryItem.sort;
    }

    private Runnable throwException(String exceptionMessage) {
        return () -> {
            throw new IllegalStateException(exceptionMessage);
        };
    }

}

File: core/src/nl/t64/game/rpg/components/party/InventoryContainer.kt
package nl.t64.game.rpg.components.party


private const val NUMBER_OF_SLOTS = 66
private const val SORTING_SPLIT = 70000 // atm, item.json starts with this number.

class InventoryContainer(numberOfSlots: Int = NUMBER_OF_SLOTS) {

    private val inventory: MutableList<InventoryItem?> = arrayOfNulls<InventoryItem>(numberOfSlots).toMutableList()

    fun getAllContent(): Map<String, Int> {
        return inventory
            .filterNotNull()
            .associate { Pair(it.id, it.amount) }   // .map(Pair(etc)).toMap()
            .toMutableMap()
    }

    fun getAllFilledSlots(): List<InventoryItem> {
        return inventory.filterNotNull()
    }

    fun getAmountOfItemAt(index: Int): Int {
        return inventory[index]?.amount ?: 0
    }

    fun incrementAmountAt(index: Int, amount: Int) {
        inventory[index]?.increaseAmountWith(amount)
            ?: throw IllegalStateException("There is no item to increment amount.")
    }

    fun decrementAmountAt(index: Int, amount: Int) {
        inventory[index]?.decreaseAmountWith(amount)
            ?: throw IllegalStateException("There is no item to decrement amount.")
    }

    fun getItemAt(index: Int): InventoryItem? {
        return inventory[index]
    }

    fun autoSetItem(newItem: InventoryItem) {
        if (newItem.isStackable) {
            addResource(newItem)
        } else {
            addItemAtEmptySlot(newItem)
        }
    }

    fun autoRemoveItem(itemId: String, orgAmount: Int) {
        check(hasEnoughOfItem(itemId, orgAmount)) { "Cannot remove this resource from Inventory." }

        var amount = orgAmount
        for ((key, value) in findAllSlotsWithAmountOfItem(itemId)) {
            if (amount == 0) {
                break
            } else if (amount >= value) {
                amount -= value
                clearItemAt(key)
            } else {
                decrementAmountAt(key, amount)
                amount = 0
            }
        }
    }

    fun clearItemAt(index: Int) {
        forceSetItemAt(index, null)
    }

    fun forceSetItemAt(index: Int, newItem: InventoryItem?) {
        inventory[index] = newItem
    }

    fun getSize(): Int {
        return inventory.size
    }

    fun isEmpty(): Boolean {
        return inventory.all { it == null }
    }

    fun sort() {
        InventoryStacksMerger(this).searchAll()
        val sortedList = inventory.sortedWith(compareBy(this::getSort, this::getInventoryGroup))
        inventory.clear()
        inventory.addAll(sortedList)
    }

    fun contains(items: Map<String, Int>): Boolean {
        return if (items.isEmpty()) {
            false
        } else {
            items.entries.all { hasEnoughOfItem(it.key, it.value) }
        }
    }

    fun hasEnoughOfItem(itemId: String, amount: Int): Boolean {
        return getTotalOfItem(itemId) >= amount
    }

    fun hasEmptySlot(): Boolean {
        return findFirstEmptySlot() != null
    }

    fun hasRoomForResource(itemId: String): Boolean {
        return (findFirstSlotWithItem(itemId) != null
                || findFirstEmptySlot() != null)
    }

    fun findFirstSlotWithItem(itemId: String): Int? {
        return (0 until getSize()).firstOrNull { containsItemAt(it, itemId) }
    }

    fun findFirstFilledSlot(): Int? {
        return findNextFilledSlotFrom(0)
    }

    fun findNextFilledSlotFrom(index: Int): Int? {
        return (index until getSize()).firstOrNull { isSlotFilled(it) }
    }

    fun findFirstEmptySlot(): Int? {
        return (0 until getSize()).firstOrNull { isSlotEmpty(it) }
    }

    fun getLastIndex(): Int {
        return getSize() - 1
    }

    fun contains(itemId: String): Boolean {
        return inventory.filterNotNull().any { it.hasSameIdAs(itemId) }
    }

    fun getTotalOfItem(itemId: String): Int {
        return inventory
            .filterNotNull()
            .filter { it.hasSameIdAs(itemId) }
            .sumOf { it.amount }
    }

    fun getNumberOfFilledSlots(): Int {
        return inventory.filterNotNull().count()
    }

    private fun addResource(newItem: InventoryItem) {
        getItem(newItem.id)?.increaseAmountWith(newItem) ?: addItemAtEmptySlot(newItem)
    }

    private fun addItemAtEmptySlot(newItem: InventoryItem) {
        findFirstEmptySlot()?.also { slotIsEmptySoSetItemAt(it, newItem) }
            ?: throw IllegalStateException("Inventory is full.")
    }

    private fun slotIsEmptySoSetItemAt(index: Int, newItem: InventoryItem) {
        forceSetItemAt(index, newItem)
    }

    private fun getItem(itemId: String): InventoryItem? {
        return inventory.filterNotNull().firstOrNull { it.hasSameIdAs(itemId) }
    }

    private fun findAllSlotsWithAmountOfItem(itemId: String): Map<Int, Int> {
        val resourceMap: MutableMap<Int, Int> = HashMap()
        (0 until getSize()).forEach { possibleAddToAmountToResourceMap(it, resourceMap, itemId) }
        return resourceMap
    }

    private fun possibleAddToAmountToResourceMap(index: Int, resourceMap: MutableMap<Int, Int>, itemId: String) {
        inventory[index]?.takeIf { it.id == itemId }?.also { resourceMap[index] = it.amount }
    }

    private fun isSlotFilled(index: Int): Boolean {
        return !isSlotEmpty(index)
    }

    private fun isSlotEmpty(index: Int): Boolean {
        return inventory[index] == null
    }

    private fun containsItemAt(index: Int, itemId: String): Boolean {
        return inventory[index]?.hasSameIdAs(itemId) ?: false
    }

    private fun getInventoryGroup(inventoryItem: InventoryItem?): InventoryGroup {
        return inventoryItem?.group ?: InventoryGroup.EMPTY
    }

    private fun getSort(inventoryItem: InventoryItem?): Int {
        return inventoryItem?.sort ?: SORTING_SPLIT
    }

}
